internal/aws/ssm_remote_access_strategy.go

```diff
@@ -149,64 +248,145 @@ func (s *SSMRemoteAccessStrategy) CleanupSSMManagedNodes(ctx context.Context, po
 	return nil
 }
 
-// isSSMRegistrationCompleted checks if SSM registration is already done for this pod
-func (s *SSMRemoteAccessStrategy) isSSMRegistrationCompleted(ctx context.Context, pod *corev1.Pod) bool {
+// isContainerRunning checks if a container is in running state
+func (s *SSMRemoteAccessStrategy) isContainerRunning(pod *corev1.Pod, containerName string) bool {
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == containerName {
+			return cs.State.Running != nil
+		}
+	}
+	return false
+}
+
+// getCurrentRestartCounts extracts restart counts from pod status
+func (s *SSMRemoteAccessStrategy) getCurrentRestartCounts(pod *corev1.Pod) (sidecarCount, workspaceCount int32) {
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == SSMAgentSidecarContainerName {
+			sidecarCount = cs.RestartCount
+		}
+		if cs.Name == WorkspaceContainerName {
+			workspaceCount = cs.RestartCount
+		}
+	}
+	return sidecarCount, workspaceCount
+}
+
+// readRegistrationState reads state from shared volume
+// Returns:
+//   - (*RegistrationState, nil) if state file exists and is valid
+//   - (nil, nil) if state file doesn't exist (first time setup)
+//   - (nil, error) if state file exists but is corrupted
+func (s *SSMRemoteAccessStrategy) readRegistrationState(ctx context.Context, pod *corev1.Pod) (*RegistrationState, error) {
+	logger := logf.FromContext(ctx).WithValues("pod", pod.Name)
+
+	cmd := []string{"cat", SSMRegistrationStateFile}
```

Comment:
This is really brittle.

Medium term: we should define a contract with the image, and hopefully ping an HTTP endpoint or trigger a pre-set `.sh` file that returns the registration state.